### ISSUE TITLE
strange behavior when begin transaction when 'tidb_snapshot' is set

### ISSUE THREAD

**Issue by lcwangchao** [pingcap/tidb#25176]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

s1, and s2 are two tidb connections.

| s1 | s2 | 
| ------ | ------ | 
| create table test.ttt (id int primary key, a int); |  | 
|  insert into test.ttt values(1, 1); | |
| | begin; |
| update test.ttt set a=2 where id=1; | |
| | do sleep(1);set @@tidb_snapshot=TIMESTAMP(NOW()); |
| | select a from test.ttt where id=1; |
| | select a from test.ttt where id=1 for update;   |
| | select a from test.ttt where id=1;   |

### 2. What did you expect to see? (Required)
```
mysql> select a from test.ttt where id=1;
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)
```

```
mysql> select a from test.ttt where id=1 for update;
+------+
| a    |
+------+
|    2 |
+------+
1 row in set (0.00 sec)
```

```
mysql> select a from test.ttt where id=1;
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)
```
mysql> select a from test.ttt where id=1;
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)
```

```
mysql> select a from test.ttt where id=1 for update;
+------+
| a    |
+------+
|    2 |
+------+
1 row in set (0.00 sec)
```

```
mysql> select a from test.ttt where id=1;
+------+
| a    |
+------+
|    2 |
+------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

master

**Comment by lcwangchao** [comment]
@cfzjywxk PTAL

**Comment by lcwangchao** [comment]
It seems when building a executor, getSnapshotTs's return uses sys var 'SnapshotTS' once it isset ignoring the txn's startTs.

https://github.com/pingcap/tidb/blob/fbbada3f0f3c2c18f8af62dbe04c93144d6d83cb/executor/builder.go#L1397-L1404

And when opening PointGetExecutor, it will step to line 135 if for update read ts is bigger than txn read ts, thats why it only happends after a select ... for update clause.

https://github.com/pingcap/tidb/blob/fbbada3f0f3c2c18f8af62dbe04c93144d6d83cb/executor/point_get.go#L132-L136

**Comment by cfzjywxk** [comment]
> It seems when building a executor, getSnapshotTs's return uses sys var 'SnapshotTS' once it isset ignoring the txn's startTs.
> 
> https://github.com/pingcap/tidb/blob/fbbada3f0f3c2c18f8af62dbe04c93144d6d83cb/executor/builder.go#L1397-L1404
> 
> And when opening PointGetExecutor, it will step to line 135 if for update read ts is bigger than txn read ts, thats why it only happends after a select ... for update clause.
> 
> https://github.com/pingcap/tidb/blob/fbbada3f0f3c2c18f8af62dbe04c93144d6d83cb/executor/point_get.go#L132-L136

@lcwangchao 
Thanks for the feedback, the behaviour is strange and we'll disscuss about the `tidb_snapshot` variable when it's used with the pessimistic transaction.

**Comment by cfzjywxk** [comment]
/cc 
@jackysp @MyonKeminta 
We may need to discuss about this behaviour setting snapshot ts during transaction execution. For pessimistic transactions, it the snapshot is set and other features like read-committed is used, I'm afraid there could be unexpected behaviours or possbile risks.

**Comment by jackysp** [comment]
I think we should narrow down the scope of tidb snapshot.

**Comment by MyonKeminta** [comment]
I agree to limit the usage of this variable. IMHO it should never be used with transactions. But limiting its usage can potentially cause compatibility issues between new and old versions of TiDB.
